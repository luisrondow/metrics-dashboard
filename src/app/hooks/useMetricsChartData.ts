import { useCallback, useMemo, useState } from "react";
import useFetchMetrics from "./useFetchMetrics";
import { convertTime } from "../utils/time";
import { TimeMeasurement } from "../types/time-utils";

const FULLY_UNPRODUCTIVE_METRIC_IDS = ["mech_problems", "unexplained", "cln_shift"]

const DEFAULT_TIME_MEASUREMENT = "secs";

export default function useMetricsChartData() {
  const { metrics: fetchedMetrics } = useFetchMetrics();

  const [timeMeasurement, setTimeMeasurement] = useState<TimeMeasurement>(DEFAULT_TIME_MEASUREMENT);

  const metrics = useMemo(() => {
    if (!fetchedMetrics) {
      return [];
    }

    return fetchedMetrics.map((metric) => {
      if ((metric.type === "secs" || metric.type === "hours") && metric.type !== timeMeasurement) {
        return {
          ...metric,
          value: Number(convertTime(metric.value, metric.type, timeMeasurement)),
          type: timeMeasurement,
        };
      }

      return metric;
    });
  }, [fetchedMetrics, timeMeasurement]);

  const { fullyUnproductive, downtime, overallEfficiency, shiftDuration } = useMemo(() => {
    let shiftDurationMetric = null;
    let overallEfficiency = null;

    const fullyUnproductive = [];
    const downtime = [];

    for (const metric of metrics) {
      if (FULLY_UNPRODUCTIVE_METRIC_IDS.includes(metric.id)) {
        fullyUnproductive.push(metric);
      }
      if (metric.category === "downtime") {
        downtime.push(metric);
      }
      if (metric.id === "shift_duration") {
        shiftDurationMetric = metric;
      }

      if (metric.id === "oee") {
        overallEfficiency = metric;
      }
    }

    return {
      fullyUnproductive,
      downtime,
      overallEfficiency,
      shiftDuration: shiftDurationMetric?.value ?? 0,
    };
  }, [metrics]);

  const efficiencyDrop = useMemo(() => {
    const efficiencyData = fullyUnproductive.reduce((acc, element) => {
      const percentageLoss = (element.value / shiftDuration) * 100;

      acc.push({
        x: element.id,
        y: Number((acc[acc.length - 1].y - percentageLoss).toFixed(1)),
      });

      return acc
    }, [
      {
        x: 'start',
        y: 100,
      },
    ]);

    return [
      {
        id: 'efficiency',
        data: [
          ...efficiencyData,
        ],
      },
    ];
  }, [fullyUnproductive, shiftDuration]);

  const lastAndAverageEfficiency = useMemo(() => {
    const averageEfficiency = (overallEfficiency?.value ?? 0) * 100;
    const lastEfficiency = efficiencyDrop[0].data[efficiencyDrop[0].data.length - 1].y;

    return [
      {
        metric: 'last shift',
        efficiency: lastEfficiency,
      },
      {
        metric: 'average shift',
        efficiency: averageEfficiency,
      },
    ];
  }, [overallEfficiency, efficiencyDrop]);

  const calculateLoss = useCallback((id: string, metricName: string) => {
    const metricValue = metrics.find((metric) => metric.id === id)?.value ?? 0;
    const efficiencyLoss = 1 - (overallEfficiency?.value ?? 0);
    const gain = Number((metricValue / efficiencyLoss).toFixed(2)) * -1;

    return [
      {
        metric: metricName,
        loss: metricValue,
        gain,
      },
    ];
  }, [metrics, overallEfficiency?.value]);

  const speedBalanceLoss = useMemo(() => calculateLoss('sl', 'Speed balance'), [calculateLoss]);
  const goodsBeforePalletizingLoss = useMemo(() => calculateLoss('lbp', 'Goods before pallets'), [calculateLoss]);

  return {
    fullyUnproductive,
    downtime,
    lastAndAverageEfficiency,
    efficiencyDrop,
    speedBalanceLoss,
    goodsBeforePalletizingLoss,
    timeMeasurement,
    setTimeMeasurement,
  };
}