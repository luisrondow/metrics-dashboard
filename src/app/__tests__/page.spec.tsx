import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Home from '../page';

import useMetricsChartData from '../hooks/useMetricsChartData';

jest.mock('../components/Header', () => {
  const Header = () => <div data-testid="header">Header</div>;
  Header.displayName = 'Header';
  return Header;
});

jest.mock('../components/charts/ChartCard', () => {
  const ChartCard = ({ children, title, isLoading }: any) => (
    <div data-testid="chart-card">
      <div>{title}</div>
      {isLoading ? <div>Loading...</div> : children}
    </div>
  );
  ChartCard.displayName = 'ChartCard';
  return ChartCard;
});

jest.mock('../components/charts/Pie', () => {
  const Pie = ({ data, valueFormatter }: any) => (
    <div data-testid="pie-chart">
      {data.map((d: any, i: number) => (
        <div key={i}>{valueFormatter(d.value)}</div>
      ))}
    </div>
  );
  Pie.displayName = 'Pie';
  return Pie;
});

jest.mock('../components/Section', () => {
  const Section = ({ children, title, headerSlot }: any) => (
    <div data-testid="section">
      <h2>{title}</h2>
      {headerSlot}
      {children}
    </div>
  );
  Section.displayName = 'Section';
  return Section;
});

jest.mock('../components/charts/Bars', () => {
  const Bars = ({ data, valueFormatter }: any) => (
    <div data-testid="bars-chart">
      {data.map((d: any, i: number) => (
        <div key={i}>{valueFormatter ? valueFormatter(d.value) : d.value}</div>
      ))}
    </div>
  );
  Bars.displayName = 'Bars';
  return Bars;
});

jest.mock('../components/charts/Line', () => {
  const Line = ({ data }: any) => (
    <div data-testid="line-chart">
      {data.map((d: any, i: number) => (
        <div key={i}>{`x: ${d.x}, y: ${d.y}`}</div>
      ))}
    </div>
  );
  Line.displayName = 'Line';
  return Line;
});

jest.mock('../components/TimeToggle', () => {
  const TimeToggle = ({ timeMeasurement, toggleTimeMeasurement }: any) => (
    <div data-testid="time-toggle">
      <button onClick={() => toggleTimeMeasurement(timeMeasurement === 'hours' ? 'secs' : 'hours')}>Toggle Time</button>
    </div>
  );
  TimeToggle.displayName = 'TimeToggle';
  return TimeToggle;
});

jest.mock('../components/Modal', () => {
  const Modal = ({ isOpen, onClose, children }: any) =>
    isOpen ? (
      <div data-testid="modal">
        <button onClick={onClose}>Close</button>
        {children}
      </div>
    ) : null;
  Modal.displayName = 'Modal';
  return Modal;
});

jest.mock('../components/Table', () => {
  const Table = ({ columns, data }: any) => (
    <div data-testid="table">
      {data.map((row: any, i: number) => (
        <div key={i}>
          {columns.map((col: string) => (
            <div key={col}>{row[col]}</div>
          ))}
        </div>
      ))}
    </div>
  );
  Table.displayName = 'Table';
  return Table;
});

const mockUseMetricsChartData = {
  fetchedMetrics: [
    {
      id: 'oee',
      label: 'oee',
      value: 0.68,
      type: 'percentage',
      description: 'The overall equipment efficiency in %',
      category: 'efficiency',
    },
  ] as any,
  isFetching: false,
  isError: false,
  fullyUnproductive: [{ value: 10 }],
  downtime: [{ value: 20 }],
  efficiencyDrop: [{ x: 'A', y: 30 }],
  lastAndAverageEfficiency: [{ metric: 'Test', value: 40 }],
  speedBalanceLoss: [{ metric: 'Test', value: 50 }],
  goodsBeforePalletizingLoss: [{ metric: 'Test', value: 60 }],
  timeMeasurement: 'secs',
  setTimeMeasurement: jest.fn(),
} as any;

jest.mock('../hooks/useTableData', () => () => ({
  columns: ['Metric', 'Value'],
  data: [{ Metric: 'Test', Value: 100 }],
  setTableData: jest.fn(),
}));

jest.mock('../hooks/useMetricsChartData');

describe('Home Component', () => {
  beforeEach(() => {
    document.body.innerHTML = '<div id="modal-root"></div>';
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    (useMetricsChartData as jest.MockedFunction<typeof useMetricsChartData>).mockReturnValue(mockUseMetricsChartData);

    const { asFragment } = render(<Home />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the header', () => {
    (useMetricsChartData as jest.MockedFunction<typeof useMetricsChartData>).mockReturnValue(mockUseMetricsChartData);

    render(<Home />);
    expect(screen.getByTestId('header')).toBeInTheDocument();
  });

  it('renders sections with titles', () => {
    (useMetricsChartData as jest.MockedFunction<typeof useMetricsChartData>).mockReturnValue(mockUseMetricsChartData);

    render(<Home />);
    expect(screen.getByText('Equipment efficiency')).toBeInTheDocument();
    expect(screen.getByText('Downtime analysis')).toBeInTheDocument();
    expect(screen.getByText('Losses')).toBeInTheDocument();
  });

  it('renders charts within sections', () => {
    (useMetricsChartData as jest.MockedFunction<typeof useMetricsChartData>).mockReturnValue(mockUseMetricsChartData);

    render(<Home />);
    expect(screen.getAllByTestId('bars-chart').length).toBe(3);
    expect(screen.getAllByTestId('pie-chart').length).toBe(2);
  });

  it('toggles time measurement', async () => {
    (useMetricsChartData as jest.MockedFunction<typeof useMetricsChartData>).mockReturnValue(mockUseMetricsChartData);

    render(<Home />);
    const toggleButton = screen.getByText('Toggle Time');
    fireEvent.click(toggleButton);
    await waitFor(() => {
      expect(toggleButton).toBeInTheDocument();
    });
  });

  it('renders error message when isError is true', () => {
    (useMetricsChartData as jest.MockedFunction<typeof useMetricsChartData>).mockReturnValue({
      ...mockUseMetricsChartData,
      isError: true,
    });

    render(<Home />);
    expect(screen.getByText('Something went wrong!')).toBeInTheDocument();
  });
});
