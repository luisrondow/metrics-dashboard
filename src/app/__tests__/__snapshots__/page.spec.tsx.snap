// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="header"
  >
    Header
  </div>
  <main>
    <div
      class="sc-b2ca270-0 jhJwBs"
    >
      <div
        data-testid="section"
      >
        <h2>
          Equipment efficiency
        </h2>
        <div
          data-testid="chart-card"
        >
          <div>
            Last sifht efficiency drop
          </div>
          <div
            data-testid="line-chart"
          >
            <div>
              x: A, y: 30
            </div>
          </div>
        </div>
        <div
          data-testid="chart-card"
        >
          <div>
            Total Working, Cleaning Shift time and Real Efficiency time
          </div>
          <div
            data-testid="bars-chart"
          >
            <div>
              40%
            </div>
          </div>
        </div>
      </div>
      <div
        data-testid="section"
      >
        <h2>
          Downtime analysis
        </h2>
        <div
          data-testid="time-toggle"
        >
          <button>
            Toggle Time
          </button>
        </div>
        <div
          data-testid="chart-card"
        >
          <div>
            Fully unproductive time
          </div>
          <div
            data-testid="pie-chart"
          >
            <div>
              10s
            </div>
          </div>
        </div>
        <div
          data-testid="chart-card"
        >
          <div>
            Strict downtime
          </div>
          <div
            data-testid="pie-chart"
          >
            <div>
              20s
            </div>
          </div>
        </div>
      </div>
      <div
        data-testid="section"
      >
        <h2>
          Losses
        </h2>
        <div
          data-testid="chart-card"
        >
          <div>
            Equipment speed balance
          </div>
          <div
            data-testid="bars-chart"
          >
            <div>
              50
            </div>
          </div>
        </div>
        <div
          data-testid="chart-card"
        >
          <div>
            Goods produced before reaching the palletizer
          </div>
          <div
            data-testid="bars-chart"
          >
            <div>
              60
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
  <div
    id="modal-root"
  />
</DocumentFragment>
`;
